In [2]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [3]:
dataset = pd.read_excel('C://Users//S1130863//Downloads/diss/curated_data.xlsx',sheet_name = 'LB_NoOthers_Final')

In [4]:
dataset

Observation Year  Standard Week  MAXT  MINT   RH1   RH2    RF   WS   SSH  \
0                1993              1  29.1  12.8  75.9  29.6   0.0  2.1   9.6   
1                1993              2  31.1  13.1  72.6  27.3   0.0  1.7  10.2   
2                1993              3  30.7  13.1  75.3  28.3   0.0  1.8   9.9   
3                1993              4  30.3  11.4  74.0  29.6   0.0  1.7   9.9   
4                1993              5  30.5  12.7  76.3  27.1   0.0  2.2  10.0   
..                ...            ...   ...   ...   ...   ...   ...  ...   ...   
359              2000             41  32.5  21.4  88.9  51.1   8.4  1.1   6.7   
360              2000             42  31.6  21.1  86.3  49.1  12.8  1.9   6.0   
361              2000             43  33.4  17.6  79.7  33.6   0.0  1.5   9.4   
362              2000             45  31.3  16.2  83.7  38.0   0.0  2.1   9.4   
363              2000             46  30.4  13.7  79.0  37.3   0.0  2.2  10.0   

     EVP   BPH  GM   GLH  LB  YSB  LB-binary  BPH-binary  GM-binary  \
0    3.2     0   0     0   0    0          0           0          0   
1    3.2     0   0     0   0    0          0           0          0   
2    3.0     0   0     0   0    0          0           0          0   
3    3.1     0   0     0   0    0          0           0          0   
4    3.1     0   0     0   0    0          0           0          0   
..   ...   ...  ..   ...  ..  ...        ...         ...        ...   
359  2.9  1662   5  1958   8  423          1           1          1   
360  3.4  2304  17  1891   8  204          1           1          1   
361  4.7  1500   6  1358   8  143          1           1          1   
362  4.1   228  14    68   0  155          0           1          1   
363  3.8     3   1     2   0    5          0           1          1   

     GLH-binary  YSB-binary  
0             0           0  
1             0           0  
2             0           0  
3             0           0  
4             0           0  
..          ...         ...  
359           1           1  
360           1           1  
361           1           1  
362           1           1  
363           0           0  

[364 rows x 20 columns]

In [31]:
years = dataset.loc[:,'Observation Year'].unique().tolist()

In [32]:
import random
selectedyears = random.sample(years,5)
selectedyears

[1998, 2000, 1996, 1999, 1994]

In [33]:
data_train = dataset[dataset['Observation Year'].isin(selectedyears)]

In [34]:
data_test = dataset[~dataset['Observation Year'].isin(selectedyears)]

In [55]:
data_test

Observation Year  Standard Week  MaxT  MinT   RH1   RH2     RF    WS  \
52               2002              3  30.4  16.0  91.0  78.0    0.0   3.6   
53               2002              6  30.9  17.6  90.0  85.0    0.0   5.2   
54               2002              7  32.9  19.7  94.0  76.0    0.0   4.3   
55               2002              8  34.5  18.8  93.0  80.0    0.0   4.8   
56               2002              9  35.4  19.0  92.0  81.0    0.0   5.1   
57               2002             10  35.2  22.2  81.0  80.0    0.0   6.8   
58               2002             11  35.8  21.5  81.0  73.0    0.0   5.7   
59               2002             12  37.0  22.4  77.0  77.0    0.0   8.1   
60               2002             13  37.3  25.4  73.0  72.0    0.0   7.9   
61               2002             15  36.7  22.9  83.0  69.0    0.0   7.8   
62               2002             16  37.4  26.0  82.0  56.0    0.0   9.8   
63               2002             17  38.4  27.0  70.0  42.0    0.0  10.0   
64               2002             18  38.4  26.8  77.0  57.0    0.0   8.8   
65               2002             27  38.8  28.9  60.0  39.0    0.0  12.0   
108              2002             14  37.0  24.0  81.0  57.0    5.0   8.1   
109              2002             19  44.5  29.4  68.0  40.0   13.5  10.5   
110              2002             20  40.1  29.1  75.0  51.0    9.8  11.8   
111              2002             21  41.9  27.4  73.0  51.0   15.0   9.9   
112              2002             22  40.1  26.7  71.0  37.0   60.2   9.2   
113              2002             23  36.8  26.2  78.0  63.0   22.0   7.0   
114              2002             24  37.6  27.2  84.0  74.0    5.5  10.0   
115              2002             25  33.7  25.1  88.0  72.0   26.4  13.5   
116              2002             26  37.9  26.6  89.0  70.0    8.4  11.8   
117              2002             28  39.1  27.2  54.0  44.0    3.9  10.9   
118              2002             29  35.2  26.5  67.0  47.0    4.7  11.6   
119              2002             30  36.4  24.0  67.0  44.0    7.7   9.2   
120              2002             31  34.2  25.3  85.0  63.0   33.3   6.2   
121              2002             32  29.6  22.4  88.0  78.0   72.8  10.5   
122              2002             33  34.8  24.7  70.0  52.0   26.2  12.7   
154              2002             34  33.4  24.2  82.0  59.0    9.5   7.6   
207              2002              1  29.0  16.9  83.0  74.0   27.5   3.9   
208              2002              2  29.3  17.4  89.0  72.0    0.0   3.0   
209              2002              4  32.4  20.5  92.0  74.0    0.0   5.1   
210              2002              5  31.0  18.8  93.0  77.0    0.0   4.7   
211              2002             43  30.4  23.3  79.0  71.0    0.0   4.1   
212              2002             45  30.8  21.3  85.0  63.0    0.0   3.9   
213              2002             47  31.2  16.9  84.0  50.0    0.0   2.8   
214              2002             50  31.1  17.1  94.0  58.0    0.0   2.3   
215              2002             51  32.9  16.3  96.0  56.0    0.0   2.9   
216              2002             52  32.0  17.4  83.0  56.0    0.0   3.8   
248              2002             35  34.0  23.1  86.0  56.0   50.5   5.8   
249              2002             36  34.2  23.9  77.0  54.0   10.0   7.5   
250              2002             37  35.5  23.3  73.0  51.0    2.2   5.0   
251              2002             38  37.1  24.0  67.0  45.0    0.0   5.2   
252              2002             39  38.0  23.5  69.0  41.0    2.0   4.7   
253              2002             40  36.9  24.1  74.0  49.0    0.0   5.0   
254              2002             41  32.2  24.5  90.0  76.0   35.8   2.7   
255              2002             42  31.8  23.7  91.0  72.0   59.9   2.9   
256              2002             44  30.5  21.2  89.0  67.0  107.0   2.4   
257              2002             46  31.9  20.2  85.0  57.0    0.0   2.3   
258              2002             48  32.0  15.8  89.0  57.0    0.0   2.4   
2

In [163]:
X = dataset.loc[:, ['MAXT','MINT','RH1','RH2','RF','WS','SSH','EVP','BPH','GM','GLH','YSB']]
y = dataset.loc[:, ['LB-binary']]

In [35]:
X_train = data_train.loc[:, ['MAXT','MINT','RH1','RH2','RF','WS','SSH','EVP','BPH-binary','GM-binary','GLH-binary','YSB-binary']]
X_test = data_test.loc[:, ['MAXT','MINT','RH1','RH2','RF','WS','SSH','EVP','BPH-binary','GM-binary','GLH-binary','YSB-binary']]
y_train = data_train.loc[:, ['LB-binary']]
y_test = data_test.loc[:, ['LB-binary']]

In [164]:
X

MAXT  MINT   RH1   RH2    RF   WS   SSH  EVP   BPH  GM   GLH  YSB
0    29.1  12.8  75.9  29.6   0.0  2.1   9.6  3.2     0   0     0    0
1    31.1  13.1  72.6  27.3   0.0  1.7  10.2  3.2     0   0     0    0
2    30.7  13.1  75.3  28.3   0.0  1.8   9.9  3.0     0   0     0    0
3    30.3  11.4  74.0  29.6   0.0  1.7   9.9  3.1     0   0     0    0
4    30.5  12.7  76.3  27.1   0.0  2.2  10.0  3.1     0   0     0    0
..    ...   ...   ...   ...   ...  ...   ...  ...   ...  ..   ...  ...
359  32.5  21.4  88.9  51.1   8.4  1.1   6.7  2.9  1662   5  1958  423
360  31.6  21.1  86.3  49.1  12.8  1.9   6.0  3.4  2304  17  1891  204
361  33.4  17.6  79.7  33.6   0.0  1.5   9.4  4.7  1500   6  1358  143
362  31.3  16.2  83.7  38.0   0.0  2.1   9.4  4.1   228  14    68  155
363  30.4  13.7  79.0  37.3   0.0  2.2  10.0  3.8     3   1     2    5

[364 rows x 12 columns]

In [165]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.30, random_state = 1)

In [166]:
X_train

MAXT       MINT        RH1        RH2         RF         WS  \
23   29.900000  17.600000  72.300000  46.000000   0.566667   3.866667   
132  28.400000  12.800000  74.400000  31.400000   0.000000   3.600000   
222  31.500000  19.800000  84.100000  41.700000   0.000000   3.300000   
111  30.366667  14.366667  82.933333  35.466667   0.000000   1.833333   
51   28.300000  10.800000  85.000000  31.600000   0.000000   1.800000   
..         ...        ...        ...        ...        ...        ...   
203  29.800000  20.500000  87.100000  63.400000  16.200000   2.700000   
255  30.700000  22.400000  81.400000  58.300000  55.500000  12.500000   
72   28.100000  13.075000  83.350000  39.350000   3.275000   2.500000   
235  30.500000  17.500000  81.000000  37.300000   2.700000   4.600000   
37   27.900000   9.900000  84.600000  39.400000   0.000000   2.000000   

           SSH        EVP   BPH  GM  GLH  YSB  
23    9.166667  11.233333     0   0    0    0  
132   9.400000   3.300000     0   0  101   13  
222   9.000000   4.300000    70   0    0   18  
111   9.666667   3.366667     0   0    0    7  
51    8.700000   2.400000     0   0    0    0  
..         ...        ...   ...  ..  ...  ...  
203   7.100000   3.000000  1960   0  995   34  
255   3.500000   3.400000    36  68    0    0  
72    8.775000  15.025000     0   0    0    0  
235   9.400000   4.700000   180   0   40   50  
37   10.300000   3.300000     0   0    0    0  

[254 rows x 12 columns]

In [167]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)
#X_train.iloc[:,:8] = sc.fit_transform(X_train.iloc[:,:8])
#X_test.iloc[:,:8] = sc.transform(X_test.iloc[:,:8])

In [168]:
X_test

array([[-0.28768564,  0.10263122,  1.03816335, ..., -0.12610964,
        -0.16948401, -0.03390083],
       [ 1.95193427,  1.02275284, -2.56277754, ..., -0.12610964,
        -0.20326149,  0.12033853],
       [-0.53349758, -1.010074  ,  1.05784062, ..., -0.12610964,
        -0.23900523, -0.43811434],
       ...,
       [-0.4788727 , -0.56071228,  0.84139062, ..., -0.10205026,
        -0.2371092 , -0.29983077],
       [ 1.65149745,  0.53059476, -1.50020482, ..., -0.12610964,
        -0.22643523,  1.39148363],
       [ 1.37837307,  0.42360388, -1.75600936, ..., -0.12610964,
        -0.23900523, -0.44875153]])

In [169]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)

C:\Users\S1130863\Anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


LogisticRegression(random_state=0)

In [170]:
y_pred = classifier.predict(X_test)

In [171]:
y_pred

array([0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
      dtype=int64)

In [172]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[84  6]
 [12  8]]


0.8363636363636363

In [173]:
from sklearn.metrics import recall_score, precision_score
recall_score(y_test, y_pred)

0.4

In [174]:
precision_score(y_test, y_pred)

0.5714285714285714

In [175]:
from sklearn.metrics import f1_score
f1_score(y_test,y_pred)

0.47058823529411764

## Results of Logistic Regression

### achieving 58% f1 and 87% accuracy with a purely weather based model and random splitting in 70:30 ratio.

### achieving greater than 91% accuracy and 72% with a weather plus pest data and random splitting in 70:30 ratio.
### Most importantly, we get disastrous results if we pass weather and pest data without pest discretization, 47% f1 score

### achieving greater than 91% accuracy and 74% f1 with year wise splitting in 5:2 ratio and weather and pest data

## KNN with just weather based gave 82% accuracy and 49% f1 split randomly 70:30 

### KNN with weather and pests split randomly achieved 89% accuracy and 71% f1

### KNN with weather and pests and year wise gave 86% accuracy and 65% f1

In [133]:
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

C:\Users\S1130863\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:179: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


In [159]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[87  3]
 [ 8 12]]


0.9

In [160]:
from sklearn.metrics import recall_score, precision_score
recall_score(y_test, y_pred)

0.6

In [161]:
precision_score(y_test, y_pred)

0.8

In [162]:
from sklearn.metrics import f1_score
f1_score(y_test,y_pred)

0.6857142857142857

In [138]:
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state = 0)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

C:\Users\S1130863\Anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


### SVC Linear gave 87.2% accuracy and f1 61% with purely weather based
### with weather plus pest rando split 90% acc and 72% f1

In [143]:
from sklearn.svm import SVC
classifier = SVC(kernel = 'rbf', random_state = 0)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

C:\Users\S1130863\Anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


### SVC rbf gave 90% acc and 78% f1 with purely weather based
### 92% and  77% with weather plus pest random split

### SVC RBF gave 89% accuracy

In [148]:
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

C:\Users\S1130863\Anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


### Naive Bayes gave disastrous 94%
### 90% and 78% with weath plus pest random split

In [153]:
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

### Decision Tree gave disastrous results of 86.5% accuracy
### DT 91% and 75% with weather plus pest

In [158]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

C:\Users\S1130863\AppData\Local\Temp/ipykernel_54440/3949669689.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  classifier.fit(X_train, y_train)


### Random Forest gave around 86.5% accuracy.
### 90% and 69% weath plus pest randsplt